src/kernel/pgtbl.c

```diff
@@ -4,184 +4,145 @@
 #include "include/cap_ops.h"
 #include "include/liveness_tbl.h"
 #include "include/retype_tbl.h"
+#include "chal/chal_prot.h"
 
 int
 pgtbl_kmem_act(pgtbl_t pt, u32_t addr, unsigned long *kern_addr, unsigned long **pte_ret)
 {
-	struct ert_intern *pte;
-	u32_t              orig_v, new_v, accum = 0;
-
-	assert(pt);
-	assert((PGTBL_FLAG_MASK & addr) == 0);
-
-	/* get the pte */
-	pte = (struct ert_intern *)__pgtbl_lkupan((pgtbl_t)((u32_t)pt | PGTBL_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT,
-	                                          PGTBL_DEPTH, &accum);
-	if (unlikely(!pte)) return -ENOENT;
-	if (unlikely(__pgtbl_isnull(pte, 0, 0))) return -ENOENT;
+	return chal_pgtbl_kmem_act(pt, addr, kern_addr, pte_ret);
+}
 
-	orig_v = (u32_t)(pte->next);
-	if (unlikely(!(orig_v & PGTBL_COSFRAME))) return -EINVAL; /* can't activate non-frames */
-	if (unlikely(orig_v & PGTBL_COSKMEM)) return -EEXIST;     /* can't re-activate kmem frames */
-	assert(!(orig_v & PGTBL_QUIESCENCE));
+int
+tlb_quiescence_check(u64_t timestamp)
+{
+	return chal_tlb_quiescence_check(timestamp);
+}
 
-	*kern_addr = (unsigned long)chal_pa2va((paddr_t)(orig_v & PGTBL_FRAME_MASK));
-	new_v      = orig_v | PGTBL_COSKMEM;
+int
+cap_memactivate(struct captbl *ct, struct cap_pgtbl *pt, capid_t frame_cap, capid_t dest_pt, vaddr_t vaddr)
+{
+	return chal_cap_memactivate(ct, pt, frame_cap, dest_pt, vaddr);
+}
 
-	/* pa2va (value in *kern_addr) will return NULL if the page is
-	 * not kernel accessible */
-	if (unlikely(!*kern_addr)) return -EINVAL; /* cannot retype a non-kernel accessible page */
-	if (unlikely(retypetbl_kern_ref((void *)(new_v & PGTBL_FRAME_MASK)))) return -EFAULT;
+int
+pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl)
+{
+	return chal_pgtbl_activate(t, cap, capin, pgtbl, lvl);
+}
 
-	/* We keep the cos_frame entry, but mark it as COSKMEM so that
-	 * we won't use it for other kernel objects. */
-	if (unlikely(cos_cas((unsigned long *)pte, orig_v, new_v) != CAS_SUCCESS)) {
-		/* restore the ref cnt. */
-		retypetbl_deref((void *)(orig_v & PGTBL_FRAME_MASK));
-		return -ECASFAIL;
-	}
-	/* Return the pte ptr, so that we can release the page if the
-	 * kobj activation failed later. */
-	*pte_ret = (unsigned long *)pte;
+int
+pgtbl_deactivate(struct captbl *t, struct cap_captbl *dest_ct_cap, unsigned long capin,
+                 livenessid_t lid, capid_t pgtbl_cap, capid_t cosframe_addr, const int root)
+{
+	return chal_pgtbl_deactivate(t, dest_ct_cap, capin, lid, pgtbl_cap, cosframe_addr, root);
+}
 
-	return 0;
+/**
+ * this works on both kmem and regular user memory: the retypetbl_ref
+ * works on both.
```

Comment:
This comment doesn't seem to add any usefulness in terms of documentation. Or does it?

Reply:
I think that @hungry-foolish is just documenting things that were not obvious when he was learning the system.  I'm fine with that.  However, some context might be useful.  Perhaps putting `Note: this works....` just to indicate that it is a small piece of documentation.